README.md
# psdash

[![Build Status](https://travis-ci.org/Jahaja/psdash.svg?branch=master)](https://travis-ci.org/Jahaja/psdash)
[![Coverage Status](https://coveralls.io/repos/Jahaja/psdash/badge.png?branch=master)](https://coveralls.io/r/Jahaja/psdash?branch=master)

psdash is a system information web dashboard for linux using data mainly served by [psutil](https://github.com/giampaolo/psutil) - hence the name.

* [Features](#features)
* [Installation](#installation)
* [Getting started](#getting-started)
* [Configuration](#configuration)
* [Screenshots](#screenshots)
* [License](#license)

## Features

* **Overview**<br>
  Dashboard overview of the system displaying data on cpu, disks, network, users, memory, swap and network.
* **Processes**<br>
    List processes (`top` like) and view detailed process information about each process.

    Apart from a detailed process overview this is also available for each process:
    * Open files
    * Open connections
    * Memory maps
    * Child processes
    * Resource limits
* **Disks**<br>
    List info on all disks and partitions.
* **Network**<br>
    List info on all network interfaces and the current throughput.
    System-wide open connections listing with filtering. Somewhat like `netstat`.
* **Logs**<br>
    Tail and search logs.
    The logs are added by patterns (like `/var/log/*.log`) which are checked periodically to account for new or deleted files.
* **Multi-node/Cluster**
    Support for multiple agent nodes that is either specified by a config or will register themselves on start-up to a common psdash node that runs the web interface.
* **All data is updated automatically, no need to refresh**

The GUI is pretty much a modified bootstrap example as I'm no designer at all.
If you got a feel for design and like to improve the UI parts of psdash, please create a pull request with your changes.
It would be much appreciated as there's much room for improvements.

## Installation

Make sure your system is able to build Python C extensions. On Debian derivatives such as Ubuntu this should translate to installing the `build-essential` and `python-dev` packages using `apt-get`:

`# apt-get install build-essential python-dev`

And on RHEL (Fedora, CentOS etc.) distributions:
```
# yum groupinstall "Development Tools"
# yum install python-devel
```

Installation using pip:<br>
`$ pip install psdash`

Since pip [1.5.1](https://github.com/pypa/pip/issues/1423) you are forced to add `--allow-external argparse` since the latest argparse is hosted on google code.

Installation from source:<br>
`$ python setup.py install`

## Getting started

Starting psdash:<br>
`$ psdash`

Starting a psdash agent:<br>
`$ psdash -a --register-to [http|https]://[host]:[port] --register-as my-agent-node`

This will start psdash in *agent mode* and try to register the node to the main psdash node pointed to by the `--register-to` option.
An agent node will setup an RPC server rather than a webserver at the host and port specified by `-p/--port` and `-b/--bind` respectively.
The main psdash node (serving HTTP) will present a list of registered nodes that are available to switch between.

Available command-line arguments:
```
$ psdash --help
usage: psdash [-h] [-l path] [-b host] [-p port] [-d] [-a]
              [--register-to host:port] [--register-as name]

psdash [version] - system information web dashboard

optional arguments:
  -h, --help            show this help message and exit
  -l path, --log path   log files to make available for psdash. Patterns (e.g.
                        /var/log/**/*.log) are supported. This option can be
                        used multiple times.
  -b host, --bind host  host to bind to. Defaults to 0.0.0.0 (all interfaces).
  -p port, --port port  port to listen on. Defaults to 5000.
  -d, --debug           enables debug mode.
  -a, --agent           Enables agent mode. This launches a RPC server, using
                        zerorpc, on given bind host and port.
  --register-to host:port
                        The psdash node running in web mode to register this
                        agent to on start up. e.g 10.0.1.22:5000
  --register-as name    The name to register as. (This will default to the
                        node's hostname)
```

## Configuration

psdash uses the configuration handling offered by Flask.
The configuration file pointed to by the environment varible `PSDASH_CONFIG` will be read on startup.<br>
e.g: `$ PSDASH_CONFIG=/home/user/config.py psdash`

In addition to the [built-in configuration values that comes with Flask](http://flask.pocoo.org/docs/config/#builtin-configuration-values) there's a number of psdash specific ones as well:

| Name | Description |
| ---- | ----------- |
| `PSDASH_AUTH_USERNAME` | When this value and `PSDASH_AUTH_PASSWORD` is set, Basic Authentication will be enabled with the provided credentials. The username of the basic authentication |
| `PSDASH_AUTH_PASSWORD` | The password of the basic authentication |
| `PSDASH_ALLOWED_REMOTE_ADDRESSES` | If this is set, only provided ip addresses will be allowed to access psdash. Addresses is separated by a comma. e.g: `PSDASH_ALLOWED_REMOTE_ADDRESSES = "10.0.0.2, 192.29.20.2"` |
| `PSDASH_URL_PREFIX` | This can be used to make psdash serve from a non-root location. e.g: `PSDASH_URL_PREFIX = "/psdash"` would make psdash serve it's pages from /psdash |
| `PSDASH_LOG_LEVEL` | The log level set for psdash (passed in to `logging.basicConfig`). *Defaults to `logging.INFO`*. |
| `PSDASH_LOG_LEVEL` | The log format set for psdash (passed in to `logging.basicConfig`). *Defaults to `%(levelname)s | %(name)s | %(message)s`*. |
| `PSDASH_NODES` | A list of psDash agent nodes (a dict per node) to register on startup. e.g `[{'name': 'mywebnode', 'host': '10.0.0.2', 'port': 5000}]` |
| `PSDASH_NET_IO_COUNTER_INTERVAL` | The interval in seconds to update the counters used for calculating network traffic. *Defaults to 3*. |
| `PSDASH_LOGS_INTERVAL` | The interval in seconds to reapply the log patterns to make sure that file-system changes are applied (log files being created or removed). *Defaults to 60*.
| `PSDASH_REGISTER_INTERVAL` | The interval in seconds to register the agent to the host psdash node. This is done periodically to be able to determine if any node has gone away and at what time. *Defaults to 60* |
| `PSDASH_LOGS` | Log patterns to apply at startup. e.g `['/var/log/*.log']`. To override this option using the command-line use the `-l/--log` arg option. |
| `PSDASH_REGISTER_TO` | When running in agent mode, this is used to set which psdash node to register the agent node to. e.g `http://10.0.20.2:5000`. |
| `PSDASH_REGISTER_AS` | When running in agent mode, this is used to set the name to register as to the host psdash node specified by `PSDASH_REGISTER_TO`. |
| `PSDASH_HTTPS_KEYFILE` | Path to the SSL key file to use to enable starting the psdash webserver in HTTPS mode. e.g `/home/user/private.key`
| `PSDASH_HTTPS_CERTFILE` | Path to the SSL certificate file to use to enable starting the psdash webserver in HTTPS mode. e.g `/home/user/certificate.crt`
| `PSDASH_ENVIRON_WHITELIST` | If set, only the env vars in this list will be displayed with value. e.g `['HOME']`

## Screenshots

Overview:
![Overview](docs/screenshots/overview.png)
Listing processes:
![Processes](docs/screenshots/processes.png)
Overview of a process:
![Process overview](docs/screenshots/process_overview.png)
Connections of a process:
![Process connections](docs/screenshots/process_connections.png)
Environment of a process:
![Process environment](docs/screenshots/process_environment.png)
Network:
![Network](docs/screenshots/network.png)
Disks:
![Network](docs/screenshots/disks.png)
Tailing a log:
![Log](docs/screenshots/log.png)
Searching a log:
![Log search](docs/screenshots/log_search.png)

## License
Released under CC0 (Public Domain Dedication).

http://creativecommons.org/publicdomain/zero/1.0/


CHANGELOG
--- v0.6.2 [2015-12-02] ---

* [NEW] Changed gevent version from 1.0.1 to 1.0.2.

* [FIX] Fixed bug that showed incorrect disk usage percentage on the overview page.

--- v0.6.1 [2014-12-13] ---

* [NEW] Changed psutil version from 2.1.1 to 2.1.3

* [FIX] Fixed bug that caused file-based configuration options to be wrongly overriden by defaulted program args.

* [FIX] Added a fix for an unhandled error in psutil when the uid of a process is not associated with an user.

* [NEW] Added a whitelist option to make it possible to decide which env vars values to display. 

--- v0.6.0 [2014-09-24] ---

* [NEW] Major updates to the UI design.

* [NEW] Added user proceses filtering to processes page.

* [FIX] Fixed bug where the highlighted text when searching a log could be misplaced.

--- v0.5.0 [2014-09-06] ---

* [NEW] Added HTTPS support.

* [NEW] Added support for multiple hosts.

* [NEW] Changed to using the netifaces lib for reading the host ip-address. (Thanks to Leonardo Santagada)

* [NEW] Made it possible to filter on remote/local address port on the system-wide connections page.

--- v0.4.0 [2014-07-24] ---

* [NEW] Added a environment variables sub-page to the process view.

* [NEW] Updated psutil to 2.1.1

* [NEW] Added a system-wide connections table on the network page.

* [FIX] Removed accidentally committed print statement.

* [NEW] Added more tests and made some refactorings as a consequence.

* [NEW] Added Travis CI setup and test coverage.

* [NEW] Added filters to system-wide connections table

* [NEW] Some rather large refactorings in general due to added tests.

--- v0.3.0 [2014-04-15] ---

* [NEW] Added a PSDASH_URL_PREFIX configuration setting to allow psdash to be served from a sub-url rather than always on root.

* [NEW] Now updating the list of available logs every minute.

* [NEW] The list of logs is now sorted in alphabetical order.

* [FIX] Fixed bug where the view logs page would crash if a previously added log becomes unavailable.

* [FIX] Fix for Python 3 support.

--- v0.2.0 [2014-04-05] ---

* [NEW] It's now possible to pass a pattern (e.g /var/log/**/*.log) when adding log files through the arg option --log.

* [FIX] Fixed issue where file size formatted values would be displayed as zero.

* [FIX] Fixed multiple issues related to non-ascii chars.

* [FIX] Fixed issue where HTML was not escaped in the log contents.

--- v0.1.2 [2014-03-31] ---

* [FIX] Fixed issue where a non-ascii process command-line would cause the process pages to crash.

* [FIX] Fixed issue where network interface info can be defaulted to a empty dict but a key was expected to exist for sorting.

* [FIX] Fixed issue where a network interface alias would cause psdash to crash

* [FIX] Improved the Vagrantfile when using a windows host.

--- v0.1.1 [2014-03-29] ---

* [FIX] Fixed error when specifying a unreadable log file. psdash will now check that
        each log file is readable on startup.

* [FIX] Fixed issue where fetching network interface addresses did not work on 32 bit arch.

* [NEW] Added a PSDASH_ prefix on all psdash specific configuration options.

* [FIX] Added argparse to setup.py requirements for Python 2.6 compatibility


psdash/run.py

# define Python 3 compatible eventloop API

class eventloop(object):
    call_later = None

try:
    import gevent
    from gevent.monkey import patch_all
    patch_all()

    from gevent.pywsgi import WSGIServer
    eventloop.call_later = gevent.spawn_later
except ImportError:
    gevent = False
    import asyncio
    eventloop = asyncio.get_event_loop()


import locale
import argparse
import logging
import socket
import urllib
#import urllib2
from logging import getLogger
from flask import Flask
#import zerorpc
from psdash import __version__
from psdash.node import LocalNode, RemoteNode
from psdash.web import fromtimestamp


logger = getLogger('psdash.run')


class PsDashRunner(object):
    DEFAULT_LOG_INTERVAL = 60
    DEFAULT_NET_IO_COUNTER_INTERVAL = 3
    DEFAULT_REGISTER_INTERVAL = 60
    DEFAULT_BIND_HOST = '0.0.0.0'
    DEFAULT_PORT = 5000
    LOCAL_NODE = 'localhost'

    @classmethod
    def create_from_cli_args(cls):
        return cls(args=None)

    def __init__(self, config_overrides=None, args=tuple()):
        self._nodes = {}
        config = self._load_args_config(args)
        if config_overrides:
            config.update(config_overrides)
        self.app = self._create_app(config)

        self._setup_nodes()
        self._setup_logging()
        self._setup_context()

    def _get_args(cls, args):
        parser = argparse.ArgumentParser(
            description='psdash %s - system information web dashboard' % __version__
        )
        parser.add_argument(
            '-l', '--log',
            action='append',
            dest='logs',
            default=None,
            metavar='path',
            help='log files to make available for psdash. Patterns (e.g. /var/log/**/*.log) are supported. '
                 'This option can be used multiple times.'
        )
        parser.add_argument(
            '-b', '--bind',
            action='store',
            dest='bind_host',
            default=None,
            metavar='host',
            help='host to bind to. Defaults to 0.0.0.0 (all interfaces).'
        )
        parser.add_argument(
            '-p', '--port',
            action='store',
            type=int,
            dest='port',
            default=None,
            metavar='port',
            help='port to listen on. Defaults to 5000.'
        )
        parser.add_argument(
            '-d', '--debug',
            action='store_true',
            dest='debug',
            help='enables debug mode.'
        )
        parser.add_argument(
            '-a', '--agent',
            action='store_true',
            dest='agent',
            help='Enables agent mode. This launches a RPC server, using zerorpc, on given bind host and port.'
        )
        parser.add_argument(
            '--register-to',
            action='store',
            dest='register_to',
            default=None,
            metavar='host:port',
            help='The psdash node running in web mode to register this agent to on start up. e.g 10.0.1.22:5000'
        )
        parser.add_argument(
            '--register-as',
            action='store',
            dest='register_as',
            default=None,
            metavar='name',
            help='The name to register as. (This will default to the node\'s hostname)'
        )

        return parser.parse_args(args)

    def _load_args_config(self, args):
        config = {}
        for k, v in vars(self._get_args(args)).items():
            if v:
                key = 'PSDASH_%s' % k.upper() if k != 'debug' else 'DEBUG'
                config[key] = v
        return config

    def _setup_nodes(self):
        self.add_node(LocalNode())

        nodes = self.app.config.get('PSDASH_NODES', [])
        logger.info("Registering %d nodes", len(nodes))
        for n in nodes:
            self.register_node(n['name'], n['host'], int(n['port']))

    def add_node(self, node):
        self._nodes[node.get_id()] = node

    def get_local_node(self):
        return self._nodes.get(self.LOCAL_NODE)

    def get_node(self, name):
        return self._nodes.get(name)

    def get_nodes(self):
        return self._nodes

    def register_node(self, name, host, port):
        n = RemoteNode(name, host, port)
        node = self.get_node(n.get_id())
        if node:
            n = node
            logger.debug("Updating registered node %s", n.get_id())
        else:
            logger.info("Registering %s", n.get_id())
        n.update_last_registered()
        self.add_node(n)
        return n

    def _create_app(self, config=None):
        app = Flask(__name__)
        app.psdash = self
        app.config.from_envvar('PSDASH_CONFIG', silent=True)

        if config and isinstance(config, dict):
            app.config.update(config)

        self._load_allowed_remote_addresses(app)

        # If the secret key is not read from the config just set it to something.
        if not app.secret_key:
            app.secret_key = 'whatisthissourcery'
        app.add_template_filter(fromtimestamp)

        from psdash.web import webapp
        prefix = app.config.get('PSDASH_URL_PREFIX')
        if prefix:
            prefix = '/' + prefix.strip('/')
        webapp.url_prefix = prefix
        app.register_blueprint(webapp)

        return app

    def _load_allowed_remote_addresses(self, app):
        key = 'PSDASH_ALLOWED_REMOTE_ADDRESSES'
        addrs = app.config.get(key)
        if not addrs:
            return

        if isinstance(addrs, (str, unicode)):
            app.config[key] = [a.strip() for a in addrs.split(',')]

    def _setup_logging(self):
        level = self.app.config.get('PSDASH_LOG_LEVEL', logging.INFO) if not self.app.debug else logging.DEBUG
        format = self.app.config.get('PSDASH_LOG_FORMAT', '%(levelname)s | %(name)s | %(message)s')

        logging.basicConfig(
            level=level,
            format=format
        )
        logging.getLogger('werkzeug').setLevel(logging.WARNING if not self.app.debug else logging.DEBUG)
        
    def _setup_workers(self):
        net_io_interval = self.app.config.get('PSDASH_NET_IO_COUNTER_INTERVAL', self.DEFAULT_NET_IO_COUNTER_INTERVAL)
        eventloop.call_later(net_io_interval, self._net_io_counters_worker, net_io_interval)

        if 'PSDASH_LOGS' in self.app.config:
            logs_interval = self.app.config.get('PSDASH_LOGS_INTERVAL', self.DEFAULT_LOG_INTERVAL)
            eventloop.call_later(logs_interval, self._logs_worker, logs_interval)

        if self.app.config.get('PSDASH_AGENT'):
            register_interval = self.app.config.get('PSDASH_REGISTER_INTERVAL', self.DEFAULT_REGISTER_INTERVAL)
            eventloop.call_later(register_interval, self._register_agent_worker, register_interval)

    def _setup_locale(self):
        # This set locale to the user default (usually controlled by the LANG env var)
        locale.setlocale(locale.LC_ALL, '')

    def _setup_context(self):
        self.get_local_node().net_io_counters.update()
        if 'PSDASH_LOGS' in self.app.config:
            self.get_local_node().logs.add_patterns(self.app.config['PSDASH_LOGS'])

    def _logs_worker(self, sleep_interval):
        while True:
            logger.debug("Reloading logs...")
            self.get_local_node().logs.add_patterns(self.app.config['PSDASH_LOGS'])
            gevent.sleep(sleep_interval)

    def _register_agent_worker(self, sleep_interval):
        while True:
            logger.debug("Registering agent...")
            self._register_agent()
            gevent.sleep(sleep_interval)

    def _net_io_counters_worker(self, sleep_interval):
        while True:
            logger.debug("Updating net io counters...")
            self.get_local_node().net_io_counters.update()
            gevent.sleep(sleep_interval)

    def _register_agent(self):
        register_name = self.app.config.get('PSDASH_REGISTER_AS')
        if not register_name:
            register_name = socket.gethostname()

        url_args = {
            'name': register_name,
            'port': self.app.config.get('PSDASH_PORT', self.DEFAULT_PORT),
        }
        register_url = '%s/register?%s' % (self.app.config['PSDASH_REGISTER_TO'], urllib.urlencode(url_args))

        if 'PSDASH_AUTH_USERNAME' in self.app.config and 'PSDASH_AUTH_PASSWORD' in self.app.config:
            auth_handler = urllib2.HTTPBasicAuthHandler()
            auth_handler.add_password(
                realm='psDash login required',
                uri=register_url,
                user=self.app.config['PSDASH_AUTH_USERNAME'],
                passwd=self.app.config['PSDASH_AUTH_PASSWORD']
            )
            opener = urllib2.build_opener(auth_handler)
            urllib2.install_opener(opener)

        try:
            urllib2.urlopen(register_url)
        except urllib2.HTTPError as e:
            logger.error('Failed to register agent to "%s": %s', register_url, e)

    def _run_rpc(self):
        logger.info("Starting RPC server (agent mode)")

        if 'PSDASH_REGISTER_TO' in self.app.config:
            self._register_agent()

        service = self.get_local_node().get_service()
        self.server = zerorpc.Server(service)
        self.server.bind('tcp://%s:%s' % (self.app.config.get('PSDASH_BIND_HOST', self.DEFAULT_BIND_HOST),
                                          self.app.config.get('PSDASH_PORT', self.DEFAULT_PORT)))
        self.server.run()

    def _run_web(self):
        logger.info("Starting web server")
        log = 'default' if self.app.debug else None

        ssl_args = {}
        if self.app.config.get('PSDASH_HTTPS_KEYFILE') and self.app.config.get('PSDASH_HTTPS_CERTFILE'):
            ssl_args = {
                'keyfile': self.app.config.get('PSDASH_HTTPS_KEYFILE'),
                'certfile': self.app.config.get('PSDASH_HTTPS_CERTFILE')
            }

        host = self.app.config.get('PSDASH_BIND_HOST', self.DEFAULT_BIND_HOST)
        port = self.app.config.get('PSDASH_PORT', self.DEFAULT_PORT)
        if gevent:
            listen_to = (host, port)
            self.server = WSGIServer(
                listen_to,
                application=self.app,
                log=log,
                **ssl_args
            )
        else:
            from wsgiref.simple_server import make_server
            self.server = make_server(host, port, self.app)
            print("Serving with wsgiref on http://%s:%s ..." % (host, port))
        self.server.serve_forever()

    def run(self):
        logger.info('Starting psdash v%s' % __version__)

        self._setup_locale()
        self._setup_workers()

        logger.info('Listening on %s:%s',
                    self.app.config.get('PSDASH_BIND_HOST', self.DEFAULT_BIND_HOST),
                    self.app.config.get('PSDASH_PORT', self.DEFAULT_PORT))

        if self.app.config.get('PSDASH_AGENT'):
            return self._run_rpc()
        else:
            return self._run_web()


def main():
    r = PsDashRunner.create_from_cli_args()
    r.run()
    

if __name__ == '__main__':
    main()